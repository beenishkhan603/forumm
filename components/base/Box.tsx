import { Theme, useTheme } from '@libs/useTheme'
import {
  CSSProperties,
  DetailedHTMLProps,
  HTMLAttributes,
  LegacyRef,
} from 'react'
import LoadingSpinner from './LoadingSpinner'

const Box = (
  props: {
    children?: any
    loading?: boolean
    color?: keyof Theme
    show?: boolean
    ignoreTheme?: boolean
    textColour?: keyof Theme
    blur?: boolean
    innerRef?: LegacyRef<HTMLDivElement>
  } & DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>
) => {
  const { theme } = useTheme()
  const {
    children,
    loading = false,
    color,
    textColour,
    show = true,
    ignoreTheme = false,
    style,
    blur = false,
    innerRef,
    ...divProps
  } = props
  if (!show) {
    return null
  }
  return (
    <div
      {...divProps}
      style={
        ignoreTheme || style
          ? style ?? {}
          : {
            backgroundColor: color && theme ? theme[color] : undefined,
            color:
              theme && textColour
                ? theme[textColour]
                : theme.textColour ?? undefined,
          }
      }
      ref={innerRef}
    >
      {loading ? (
        <Box
          data-test-id={'box-spinner'}
          className="w-full h-full justify-center items-center flex"
        >
          <LoadingSpinner size="medium" />
        </Box>
      ) : blur ? (
        <span style={{ filter: 'blur(2px)' }}>{children}</span>
      ) : (
        children
      )}
    </div>
  )
}

export default Box
