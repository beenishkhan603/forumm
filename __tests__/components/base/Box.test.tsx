import Box from '@components/base/Box'
import { render, waitFor, screen } from '@testing-library/react'

it('renders box unchanged', () => {
  const { container } = render(<Box>Test</Box>)
  expect(container).toMatchSnapshot()
})

it('SHOULD return null WHEN show is false', () => {
  render(<Box show={false}>Test</Box>)
  waitFor(async () => {
    const testEl = await screen.findByText('Test')
    expect(testEl).not.toBeInTheDocument()
  })
})

it('SHOULD renders box with style attributes', () => {
  render(<Box loading>Test</Box>)
  waitFor(async () => {
    const spinnerEl = await screen.findByTestId('box-spinner')
    expect(spinnerEl).toBeInTheDocument()
    const testEl = await screen.findByText('Test')
    expect(testEl).not.toBeInTheDocument()
  })
})

it('SHOULD renders box with style attributes', () => {
  render(<Box loading={false}>Test</Box>)
  waitFor(async () => {
    const spinnerEl = await screen.findByTestId('box-spinner')
    expect(spinnerEl).not.toBeInTheDocument()
    const testEl = await screen.findByText('Test')
    expect(testEl).toBeInTheDocument()
  })
})

it('SHOULD renders box with style attributes', () => {
  render(<Box style={{ display: 'none' }}>Test</Box>)
  waitFor(async () => {
    const testEl = await screen.findByText('Test')
    expect(testEl).not.toBeInTheDocument()
  })
})

it('SHOULD renders box with style attributes', () => {
  render(<Box ignoreTheme>Test</Box>)
  waitFor(async () => {
    const testEl = await screen.findByText('Test')
    expect(testEl).toBeInTheDocument()
  })
})

it('SHOULD renders box with style attributes', () => {
  render(<Box color={'foregroundColour'}>Test</Box>)
  waitFor(async () => {
    const testEl = await screen.findByText('Test')
    expect(testEl).toBeInTheDocument()
    expect(testEl).toHaveStyle('')
  })
})

it('SHOULD renders box with style attributes', () => {
  render(<Box textColour={'foregroundColour'}>Test</Box>)
  waitFor(async () => {
    const testEl = await screen.findByText('Test')
    expect(testEl).toBeInTheDocument()
    expect(testEl).toHaveStyle('')
  })
})

it('SHOULD renders box with style attributes', () => {
  jest.mock('@libs/useTheme', () => ({
    useTheme() {
      return {
        theme: {
          color: undefined,
          textColor: undefined,
        },
      }
    },
  }))
  render(<Box>Test</Box>)
  waitFor(async () => {
    const testEl = await screen.findByText('Test')
    expect(testEl).toBeInTheDocument()
    expect(testEl).toHaveStyle('')
  })
})
