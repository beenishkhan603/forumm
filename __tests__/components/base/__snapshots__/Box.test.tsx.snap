// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders box unchanged 1`] = `
<div>
  <div
    style="color: rgb(24, 26, 32);"
  >
    Test
  </div>
</div>
`;
